presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -283,15 +283,15 @@ private boolean isInliningCandidate(Expression expression, ProjectNode node)
 
             // candidate symbols for inlining are
             //   1. references to simple constants
-            //   2. references to complex expressions that appear only once
+            //   2. references to complex expressions that appear <= 4
             // which come from the node, as opposed to an enclosing scope.
             Set<Symbol> childOutputSet = ImmutableSet.copyOf(node.getOutputSymbols());
             Map<Symbol, Long> dependencies = SymbolsExtractor.extractAll(expression).stream()
                     .filter(childOutputSet::contains)
                     .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
 
             return dependencies.entrySet().stream()
-                    .allMatch(entry -> entry.getValue() == 1 || node.getAssignments().get(entry.getKey()) instanceof Literal);
+                    .allMatch(entry -> entry.getValue() <= 4 || node.getAssignments().get(entry.getKey()) instanceof Literal);
```

Comment:
changing it to `4` might cause planning to explode exponentially (in fact 2 might cause it).
I think we could change entire condition to:
```
entry -> entry.getValue() <= 4 || SymbolsExtractor.extractAll(node.getAssignments().get(entry.getKey())).size() == 1
```
This way predicate won't grow exponentially but would be pushed though simple expressions that reference only single symbol

Reply:
I think it won't, because `4` here means that the biggest explosion is limited to 4.

Reply:
It *will* cause some explosion in the number of terms. E.g., for a query like:

```
    WITH
    t1 (v) AS (VALUES 1),
    t2 AS( select if(v = 0, v, v) v from t1 ),
    t3 AS( select if(v = 0, v, v) v from t2 ),
    t4 AS( select if(v = 0, v, v) v from t3 ),
    t5 AS( select if(v = 0, v, v) v from t4 ),
    t6 AS( select if(v = 0, v, v) v from t5 ),
    t7 AS( select if(v = 0, v, v) v from t6 ),
    t8 AS( select if(v = 0, v, v) v from t7 ),
    t9 AS( select if(v = 0, v, v) v from t8 ),
    t10 AS( select if(v = 0, v, v) v from t9 ),
    t11 AS( select if(v = 0, v, v) v from t10 ),
    t12 AS( select if(v = 0, v, v) v from t11 ),
    t13 AS( select if(v = 0, v, v) v from t12 ),
    t14 AS( select if(v = 0, v, v) v from t13 ),
    t15 AS( select if(v = 0, v, v) v from t14 ),
    t16 AS( select if(v = 0, v, v) v from t15 )
    select *
    from t16
    where v = 0
```

One of the filters will be:

```
((CASE 
  WHEN ((CASE WHEN ("expr_62" = 0) THEN "expr_62" ELSE "expr_62" END) = 0) 
  THEN 
    (CASE WHEN ("expr_62" = 0) THEN "expr_62" ELSE "expr_62" END) 
  ELSE 
    (CASE WHEN ("expr_62" = 0) THEN "expr_62" ELSE "expr_62" END) END
) = 0)
```

Reply:
Yes, but I think this explosion is controlled by `<= 4`. Or we can change it smaller, but in my case, `== 1` will make `OR`'s pushdown fail.
It's just a workaround I think, not a final solution of `OR`'s pushdown.